python/shotgun_model/shotgun_entity_model.py

```diff
@@ -0,0 +1,249 @@
+        :param schema_generation:   Schema generation index. If you are changing the format
+                                    of the data you are retrieving from Shotgun, and therefore
+                                    want to invalidate any cache files that may already exist
+                                    in the system, you can increment this integer.
+        :param bg_load_thumbs:      If set to True, thumbnails will be loaded in the background.
+        :param bg_task_manager:     Background task manager to use for any asynchronous work.  If
+                                    this is None then a task manager will be created as needed.
+        """
+        self._entity_icons = {}
+        self._step_swatch_icons = {}
+
+        # make sure fields is valid:
+        fields = fields or []
+        # for backwards compatibility, make sure certain fields are added:
+        fields = list(set(fields + ["image", "sg_status_list", "description"]))
+
+        ## folder icon
+        self._default_icon = QtGui.QIcon(QtGui.QPixmap(":/tk-framework-shotgunutils/icon_Folder.png"))
+
+        ShotgunModel.__init__(self, 
+                              parent = parent,
+                              download_thumbs = download_thumbs,
+                              schema_generation = schema_generation,
+                              bg_load_thumbs = bg_load_thumbs,
+                              bg_task_manager = bg_task_manager)
+
+        # load the data from the cache:
+        self._load_data(entity_type, filters, hierarchy, fields)
+
+    def destroy(self):
+        """
+        Call to clean-up the model when it is finished with
+        """
+        ShotgunModel.destroy(self)
+        self._entity_icons = {}
+        self._step_swatch_icons = {}
+        self._default_icon = None
+
+    def get_entity_icon(self, entity_type):
+        """
+        Retrieve the icon for the specified entity type if available.
+
+        :param entity_type: The entity type to retrieve the icon for
+        :returns:           A QIcon if an icon was found for the specified entity
+                            type, otherwise None.
+        """
+        icon = None
+        if entity_type in self._entity_icons:
+            # we've previously asked for the icon
+            icon = self._entity_icons[entity_type]
+        else:
+            # see if we have the icon in the resources:
+            icon_path = ":/tk-framework-shotgunutils/icon_%s_dark.png" % entity_type
+            if QtCore.QFile.exists(icon_path): 
+                # create the new icon from this resource:
+                icon = QtGui.QIcon(QtGui.QPixmap(icon_path))
+            self._entity_icons[entity_type] = icon
+
+        return QtGui.QIcon(icon) if icon else None
+
+    def get_entities(self, item):
+        """
+        Get entities for the current item by traversing up the tree and pulling 
+        entity information from each item if possible
```

Comment:
For the parent levels, what fields are returned? is it name/type/id? I think an example here would be useful, for example for a configuration with `Asset` and hierarchy `[project, asset_type, code]`